### PR TITLE
Add direct link for beatmap packs

### LINKED ISSUE #4246: Keep direct links for Beatmap Packs on new website
At the moment you can post a link directly from a beatmap pack from the old website design, the format is 
`https://osu.ppy.sh/p/packlist/?n=<letter><number>#<letter><number>` and is a pretty handy feature, it's not that much used, yet I work a lot with that.

This however is not possible for the new forum design, and I would love if this could be kept. The current URL on the new website is `https://osu.ppy.sh/beatmaps/packs<letter><number>` which is pretty easy to use it but doesn't seem to work.

### PR DESCRIPTION
**Pull request by cl8n**
fixes #4246 , external links should use /beatmaps/packs/{id} or /beatmaps/packs/{tag}

### CODE REVIEW

File: app/Http/Controllers/BeatmapPacksController.php [hunk 1]
```diff
@@ -55,4 +67,14 @@ public function show($id)
 
         return view('packs.show', compact('pack', 'sets', 'mode'));
     }
+
+    private function indexLink(BeatmapPack $pack) : string
+    {
+        $type = $pack->type();
+        $indexInPagination = BeatmapPack::getPacks($type)->get()->search($pack);
```
Comment (nanaya) [inline app/Http/Controllers/BeatmapPacksController.php]: Not sure if want to do this. It doesn't expand either. Might as well just display expanded single item.

File: app/Http/Controllers/BeatmapPacksController.php [hunk 2]
```diff
@@ -38,11 +39,22 @@ public function index()
...
+        if (is_numeric($idOrTag)) {
+            $pack = BeatmapPack::findOrFail($idOrTag);
+        } else {
+            $pack = BeatmapPack::where('tag', $idOrTag)->firstOrFail();
+        }
+
+        return ujs_redirect($this->indexLink($pack));
+    }
+
+    public function raw($id)
```
Comment (nanaya) [inline app/Http/Controllers/BeatmapPacksController.php]: the view is still called `show`.

File: resources/views/packs/index.blade.php [hunk 3]
```diff
@@ -52,6 +52,7 @@
             <div class="beatmap-packs__list js-accordion">
                 @foreach ($packs as $pack)
                     <div class="beatmap-pack js-beatmap-pack js-accordion__item" data-pack-id="{{ $pack['pack_id'] }}">
+                        <a id="pack-{{ $pack['pack_id'] }}"></a>
```
Comment (nanaya) [inline resources/views/packs/index.blade.php]: This doesn't work and blocks things above it.

**Comment by nanaya** [comment]
Also the list items should be changed to link instead of plain div.

File: app/Http/Controllers/BeatmapPacksController.php [hunk 4]
```diff
@@ -38,11 +39,22 @@ public function index()
...
         return view('packs.index')
-            ->with('packs', $packs->paginate(20)->appends(['type' => $type]))
+            ->with('packs', $packs->paginate(static::PER_PAGE)->appends(['type' => $type]))
             ->with('type', $type);
     }
 
-    public function show($id)
+    public function show($idOrTag)
+    {
+        if (is_numeric($idOrTag)) {
```
Comment (notbakaneko) [inline app/Http/Controllers/BeatmapPacksController.php]:
```php
>>> is_numeric('1e2')
=> true
>>> is_numeric('1.2')
=> true
>>> is_numeric('1')
=> true
>>> ctype_digit('1.2')
=> false
>>> ctype_digit('1e2')
=> false
>>> ctype_digit('1')
=> true
```

**Comment by notbakaneko** [comment]
Closing due to lack of updates

**Comment by peppy** [comment]
Reopening this one, should be easy to complete. @cl8n can you let us know if you plan on fixing this up? If not one of us will get it updated and merged.

**Comment by cl8n** [comment]
not in the near future

iirc I didn't continue with the PR mostly because this seemed like a dumb workaround, but I couldn't figure out a better solution (nanaya commented there too)
https://github.com/ppy/osu-web/blob/218f097379fe01cd31d2b0079d2c6d8f5e815397/app/Http/Controllers/BeatmapPacksController.php#L71-L79

**Comment by peppy** [comment]
Can someone look into this? It looks to be a prerequisite of a [news post](https://github.com/ppy/osu-wiki/pull/2667)

**Comment by notbakaneko** [comment]
I think we should just go with adding a page that can show a single pack with a link back to the index.

**Comment by peppy** [comment]
Sounds fine if that's the easiest way to resolve this.

**Comment by nanaya** [comment]
Updated to
- show single item when accessed directly
- add link to them
- reworked css classes

File: app/Models/BeatmapPack.php [hunk 5]
```diff
@@ -66,6 +66,11 @@ public function downloadUrl()
         return $this->downloadUrls()[0];
     }
 
+    public function type() : string
+    {
+        return array_search($this->tag[0], static::$tagMappings, true);
+    }
+
```
Comment (notbakaneko) [inline app/Models/BeatmapPack.php]: Don't need this, I think?

**Pull request merged**